### ISSUE TITLE
Counterintuitive Running time on GPU, compiled differences on profiler for different shapes

### ISSUE THREAD

**Issue by iolloj** [google/jax#21336]
### Description

I am running some sort of stochastic optimisation on GPU and depending on the shape of my samples I get very different running times.
For example, the following code with shapes of 100x200 yields a running time of 28.8s while for 1000x2000 I have 5.8s.

I have trouble understanding this **increase** in performance on GPU as the number of samples is x100. I expected low sample size to be somewhat inefficient on GPU but not to have its performance reduce by a factor 6 compared to 100 times more samples. Is there something I am missing ? 

A minimum reproducible example is:
```python
import jax
from jaxtyping import PRNGKeyArray, Array
from functools import partial
import jax.numpy as jnp
import optax
import time

import numpyro.distributions as dist
max_signal = 1.0
base_signal = 0.1
noise_var = 0.1

jax.print_environment_info()


def forward(xi, theta):
    inv_value = max_signal + jnp.power(theta - xi, 2).sum(-1)
    arr = base_signal + jnp.power(inv_value, -1).sum(-1)
    return arr

def log_prob(thetas, y, xi):
    f_values = jax.vmap(forward, in_axes=(0, None), out_axes=-1)(xi, thetas)
    return (
        dist.Normal(0, noise_var)
        .log_prob(jnp.log(y) - jnp.log(f_values))
        .sum(-1)
    )

def sample(args, rng_key, xi):
    f_values = jax.vmap(forward, in_axes=(0, None), out_axes=-1)(xi, args)
    return jnp.exp(
        jnp.log(f_values)
        + dist.Normal(0, noise_var).sample(
            rng_key, sample_shape=f_values.shape
        )
    )


def pce_bound(
    design: Array,
    rng_key: PRNGKeyArray,
    thetas: Array):

    # sample N y_n
    N_samples = jax.tree_map(lambda leaf: leaf[:, 0], thetas)
    N, Lpp = jax.tree_util.tree_leaves(thetas)[0].shape[:2]
    sample_keys = jax.random.split(rng_key, N)
    sampler = partial(sample, xi=design)
    N_y = jax.vmap(sampler, (0, 0))(N_samples, sample_keys)

    # corresponding NxL log p(y_n| theta_l, design)
    log_liks_nm = jax.vmap(log_prob, in_axes=(1, None, None), out_axes=1)(
        thetas, N_y, design
    )

    # (Nx1) log probs
    log_liks_n = log_liks_nm[:, 0]

    contr_log_lik = jax.scipy.special.logsumexp(log_liks_nm, axis=1) - jnp.log(Lpp)

    return jnp.mean(log_liks_n - contr_log_lik)

# Number of samples
#N, L = 200, 100
N, L = 2000, 1000
rng_key = jax.random.PRNGKey(0)
thetas = jax.random.normal(rng_key, (N, L, 2, 2))
design = jnp.array([[0.1, 0.1]])


optimizer = optax.adam(1e-3)
optimizer_state = optimizer.init(design)

def step(state, rng_key):
    design, optimizer_state = state
    grad = jax.grad(pce_bound)(design, rng_key, thetas)
    updates, optimizer_state = optimizer.update(grad, optimizer_state)
    design = optax.apply_updates(design, updates)
    return (design, optimizer_state), design

start = time.time()
keys = jax.random.split(rng_key, 5000)
design, _ = jax.lax.scan(step, (design, optimizer_state), keys)
end = time.time()
print(end - start)
```

I investigated a bit with the profiler and the main difference is that the code seems to be ran  differently. For low sample sizes, we get the function add_any that takes more time to run compared to its reduce_sum counterpart. 
Is there something I am missing that would explain such differences?

<img width="737" alt="Screenshot 2024-01-25 at 13 21 39" src="https://github.com/google/jax/assets/48028150/47a73698-597c-4ff0-bff7-e63b803e90fb">
<img width="736" alt="Screenshot 2024-01-25 at 13 22 14" src="https://github.com/google/jax/assets/48028150/19bf001d-f384-43b7-82f9-ab0b8a0abb5f">

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.4.23
jaxlib: 0.4.23
numpy:  1.26.3
python: 3.10.10 (main, Mar 21 2023, 18:45:11) [GCC 11.2.0]
jax.devices (1 total, 1 local): [cuda(id=0)]
process_count: 1

$ nvidia-smi
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.129.03             Driver Version: 535.129.03   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla V100-SXM2-32GB           On  | 00000000:8A:00.0 Off |                    0 |
| N/A   51C    P0              76W / 300W |    311MiB / 32768MiB |      2%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A     60171      C   ...d68za/.conda/envs/jax11/bin/python3      308MiB |
+---------------------------------------------------------------------------------------+

**Comment by iolloj** [comment]
Hi, thanks for your response. I believe this difference is neither due to compilation time nor asynchronous dispatch as I get same results with jax.block_until_ready() and compilation time seems to be <1s.

**Comment by jakevdp** [comment]
Your updated script doesn't seem to isolate compilation time – even if you're not using `jax.jit` yourself, it is used internally by APIs you are calling.

**Comment by iolloj** [comment]
Yes - compilation happens inside scan. To asses compilation time I compared two consecutive calls to timer block:
```python
start = time.time()
keys = jax.random.split(rng_key, 5000)
design, _ = jax.lax.scan(step, (design, optimizer_state), keys)
jax.block_until_ready(design)
end = time.time()
print(end - start) # compilation happens 6s

start = time.time()
keys = jax.random.split(rng_key, 5000)
design, _ = jax.lax.scan(step, (design, optimizer_state), keys)
jax.block_until_ready(design)
end = time.time()
print(end - start) # no compilation 5.2s
```
which suggest that compilation takes <1s.

My question was regarding the potential differences between XLA reduce_sum and add_any and why the same code with different shapes gets compiled in two more or less efficient ways. The most efficient way being the one compiled for the largest shapes.

On top of empirical total running times, the difference between the two can be assessed with a tensorboard profiler for example where we actually see the difference between add_any / reduce_sum. The function step() takes longer to run in the add_any case even though there are less computations to do. 

This behaviour is not observed on CPU where running time is proportional to shapes (2.5s for 100x200 and 126s for 1000x2000).
